Mathlib/MeasureTheory/Measure/Haar/Quotient.lean

```diff
@@ -6,159 +6,380 @@ Authors: Alex Kontorovich, Heather Macbeth
+  invariant measure `ν` on `G`, then it is a `G` invariant measure on `G ⧸ Γ`. -/
+@[to_additive]
+lemma MeasureTheory.QuotientMeasureEqMeasurePreimage.smulInvariantMeasure_quotient
+    [IsMulLeftInvariant ν] [hasFun : HasFundamentalDomain Γ.op G ν] :
+    SMulInvariantMeasure G (G ⧸ Γ) μ where
   measure_preimage_smul g A hA := by
-    let π : G → G ⧸ Γ := QuotientGroup.mk
     have meas_π : Measurable π := continuous_quotient_mk'.measurable
-    have 𝓕meas : NullMeasurableSet 𝓕 μ := h𝓕.nullMeasurableSet
-    have meas_πA : MeasurableSet (π ⁻¹' A) := measurableSet_preimage meas_π hA
-    rw [Measure.map_apply meas_π hA,
-      Measure.map_apply meas_π (measurableSet_preimage (measurable_const_smul g) hA),
-      Measure.restrict_apply₀' 𝓕meas, Measure.restrict_apply₀' 𝓕meas]
+    obtain ⟨𝓕, h𝓕⟩ := hasFun.ExistsIsFundamentalDomain
+    have h𝓕_translate_fundom : IsFundamentalDomain Γ.op (g • 𝓕) ν :=
+      h𝓕.smul_of_comm g
+    rw [h𝓕.projection_respects_measure_apply (μ := μ)
+      (meas_π (measurableSet_preimage (measurable_const_smul g) hA)),
+      h𝓕_translate_fundom.projection_respects_measure_apply (μ := μ) hA]
+    change ν ((π ⁻¹' _) ∩ _) = ν ((π ⁻¹' _) ∩ _)
     set π_preA := π ⁻¹' A
-    have : π ⁻¹' ((fun x : G ⧸ Γ => g • x) ⁻¹' A) = (g * ·) ⁻¹' π_preA := by
-      ext1; simp
+    have : π ⁻¹' ((fun x : G ⧸ Γ => g • x) ⁻¹' A) = (g * ·) ⁻¹' π_preA := by ext1; simp
     rw [this]
-    have : μ ((g * ·) ⁻¹' π_preA ∩ 𝓕) = μ (π_preA ∩ (g⁻¹ * ·) ⁻¹' 𝓕) := by
-      trans μ ((g * ·) ⁻¹' (π_preA ∩ (g⁻¹ * ·) ⁻¹' 𝓕))
+    have : ν ((g * ·) ⁻¹' π_preA ∩ 𝓕) = ν (π_preA ∩ (g⁻¹ * ·) ⁻¹' 𝓕)
+    · trans ν ((g * ·) ⁻¹' (π_preA ∩ (g⁻¹ * ·) ⁻¹' 𝓕))
       · rw [preimage_inter]
         congr 2
         simp [Set.preimage]
       rw [measure_preimage_mul]
-    rw [this]
-    have h𝓕_translate_fundom : IsFundamentalDomain Γ.op (g • 𝓕) μ :=
-      h𝓕.smul_of_comm g
-    rw [h𝓕.measure_set_eq h𝓕_translate_fundom meas_πA, ← preimage_smul_inv]; rfl
-    rintro ⟨γ, γ_in_Γ⟩
-    ext x
-    have : π (x * MulOpposite.unop γ) = π x := by simpa [QuotientGroup.eq'] using γ_in_Γ
-    simp only [(· • ·), ← this, mem_preimage]
+    rw [this, ← preimage_smul_inv]; rfl
+
+/-- Given a subgroup `Γ` of a topological group `G` with measure `ν`, and a measure 'μ' on the
+  quotient `G ⧸ Γ` satisfying `QuotientMeasureEqMeasurePreimage`, the restriction
+  of `ν` to a fundamental domain is measure-preserving with respect to `μ`. -/
+@[to_additive measurePreserving_addQuotientGroup_mk_of_addQuotientMeasureEqMeasurePreimage]
+theorem measurePreserving_quotientGroup_mk_of_QuotientMeasureEqMeasurePreimage
+    {𝓕 : Set G} (h𝓕 : IsFundamentalDomain Γ.op 𝓕 ν) (μ : Measure (G ⧸ Γ))
+    [QuotientMeasureEqMeasurePreimage ν μ] :
+    MeasurePreserving (@QuotientGroup.mk G _ Γ) (ν.restrict 𝓕) μ :=
+  h𝓕.measurePreserving_quotient_mk ν μ
+
+end smulInvariantMeasure
+
+section normal
+
+section additive
```

Comment:
I am confused by the need to have a section on additive group. Can't everything be proved on multiplicative groups, and then transferred by `to_additive`? I checked and indeed a proof below breaks, but you can replace the line `convert one_smul ENNReal μ` with `suffices (μ' V / ν (QuotientGroup.mk ⁻¹' V ∩ s)) = 1 by rw [this, one_smul]` to make `to_additive` happy. (`to_additive` is often confused by `convert` and `congr`).

Reply:
Great, thanks!